File: arenagymapk/lib/models/meta.dart
class Meta {
  final int idMeta;
  final int idUsuario;
  final String tipoMeta;
  final double? valorInicial;
  final double? valorObjetivo;
  final double? valorActual;
  final String fechaInicio;
  final String? fechaObjetivo;
  final String estado;
  final String? descripcion;

  Meta({
    required this.idMeta,
    required this.idUsuario,
    required this.tipoMeta,
    required this.valorInicial,
    required this.valorObjetivo,
    required this.valorActual,
    required this.fechaInicio,
    required this.fechaObjetivo,
    required this.estado,
    required this.descripcion,
  });

  /// Progreso de 0.0 a 1.0 entre [valorInicial] y [valorObjetivo], según
  /// dónde se ubica [valorActual]. Null si no hay suficientes datos.
  double? get progreso {
    if (valorInicial == null || valorObjetivo == null || valorActual == null) {
      return null;
    }
    final rango = valorObjetivo! - valorInicial!;
    if (rango == 0) return 1.0;
    final avance = (valorActual! - valorInicial!) / rango;
    return avance.clamp(0.0, 1.0);
  }

  factory Meta.fromJson(Map<String, dynamic> json) {
    return Meta(
      idMeta: json['id_meta'] as int,
      idUsuario: json['id_usuario'] as int? ?? 0,
      tipoMeta: json['tipo_meta']?.toString() ?? '',
      valorInicial: double.tryParse(json['valor_inicial']?.toString() ?? ''),
      valorObjetivo: double.tryParse(json['valor_objetivo']?.toString() ?? ''),
      valorActual: double.tryParse(json['valor_actual']?.toString() ?? ''),
      fechaInicio: json['fecha_inicio']?.toString() ?? '',
      fechaObjetivo: json['fecha_objetivo'] as String?,
      estado: json['estado']?.toString() ?? 'activa',
      descripcion: json['descripcion'] as String?,
    );
  }
}

File: frontend-gym/src/lib/db/models/meta.js
const { DataTypes } = require('sequelize');
const sequelize = require('../sequelize');

const Meta = sequelize.define('Meta', {
  id_meta: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true,
  },
  id_usuario: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: 'usuarios',
      key: 'id_usuario'
    }
  },
  tipo_meta: {
    type: DataTypes.STRING(50),
    allowNull: false
  },
  valor_inicial: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: true
  },
  valor_objetivo: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: true
  },
  valor_actual: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: true
  },
  fecha_inicio: {
    type: DataTypes.DATEONLY,
    defaultValue: DataTypes.NOW
  },
  fecha_objetivo: {
    type: DataTypes.DATEONLY,
    allowNull: true
  },
  estado: {
    type: DataTypes.STRING(20),
    allowNull: false,
    defaultValue: 'activa',
    validate: {
      isIn: [['activa', 'cumplida', 'cancelada']]
    }
  },
  descripcion: {
    type: DataTypes.TEXT,
    allowNull: true
  }
}, {
  tableName: 'metas',
  timestamps: false,
});

module.exports = Meta;
